pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,430 @@
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest
+	/// payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Seconds) -> Result<bool, DispatchError> {
+		match self.interest_payments {
+			InterestPayments::None => (),
+			InterestPayments::Monthly(_) => {
+				let start = date::from_seconds(now)?;
+				let end = date::from_seconds(self.maturity.date())?;
+
+				// We want to avoid creating a loan with a cashflow consuming a lot of computing
+				// time Maximum 40 years, which means a cashflow list of 40 * 12 elements
+				if end.year() - start.year() > 40 {
+					return Ok(false);
+				}
+			}
+		}
+
+		Ok(self.maturity.is_valid(now))
+	}
+
+	pub fn generate_cashflows<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+	) -> Result<Vec<CashflowPayment<Balance>>, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber + EnsureDiv,
+	{
+		let start_date = date::from_seconds(origination_date)?;
+		let end_date = date::from_seconds(self.maturity.date())?;
+
+		let (timeflow, periods_per_year) = match &self.interest_payments {
+			InterestPayments::None => (vec![], 1),
+			InterestPayments::Monthly(reference_day) => (
+				date::monthly_intervals(start_date, end_date, (*reference_day).into())?,
+				12,
+			),
+		};
+
+		let total_weight = timeflow
+			.iter()
+			.map(|(_, weight)| weight)
+			.try_fold(0, |a, b| a.ensure_add(*b))?;
+
+		let interest_per_period = interest_rate
+			.per_year()
+			.ensure_div(Rate::saturating_from_integer(periods_per_year))?;
+
+		timeflow
+			.into_iter()
+			.map(|(date, weight)| {
+				let proportion = FixedU128::ensure_from_rational(weight, total_weight)?;
+				let principal = proportion.ensure_mul_int(principal)?;
+				let interest = interest_per_period.ensure_mul_int(principal)?;
+
+				Ok(CashflowPayment {
+					when: date::into_seconds(date)?,
+					principal,
+					interest,
+				})
+			})
+			.collect()
+	}
+
+	pub fn expected_payment<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+		until: Seconds,
+	) -> Result<Balance, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber,
+	{
+		let cashflow = self.generate_cashflows(origination_date, principal, interest_rate)?;
+
+		let total_amount = cashflow
+			.into_iter()
+			.take_while(|payment| payment.when < until)
+			.map(|payment| payment.principal.ensure_add(payment.interest))
+			.try_fold(Balance::zero(), |a, b| a.ensure_add(b?))?;
+
+		Ok(total_amount)
+	}
```

Comment:
> not more interest or principal then expected is repayed.

Why this restriction? Could not anyone pay in advance for the upcoming months if they want?

> Currently, we are only checking if outstanding repayments are overdue when borrowing again.

Yes, this is the MVP of this feature I talked with @hieronx about. We're only checking that to prevent the cashflow history from being broken.

Reply:
> If you are repaying the principal, you are decreasing the future cashflows

Yeah, ideally, the repayment should be amortized. By now, if you repay more, it means that you do not need to repay the next month. With 0 amortizations.